src/Examine/LuceneEngine/Directories/ExamineDirectory.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Lucene.Net.Index;
+using static Lucene.Net.Index.IndexWriter;
+
+namespace Examine.LuceneEngine.Directories
+{
+    public abstract class ExamineDirectory : Lucene.Net.Store.Directory
+    {
+        private Action<ExamineIndexWriter> _onCommitAction;
+        public virtual void InvokeOnCommit(ExamineIndexWriter writer)
+        {
+            _onCommitAction?.Invoke(writer);
+        }
+
+        public event EventHandler HandleOutOfSyncEvent;
+
+        public void HandleOutOfSyncDirectory()
+        {
+            HandleOutOfSyncEvent?.Invoke(this, null);
+        }
+
+        /// <summary>
+        /// Called on commit
+        /// </summary>
+        /// <param name="action"></param>
+        public void SetOnCommitAction(Action<ExamineIndexWriter> action)
```

Comment:
All of these `Set` methods mean that you can mutate this instance at runtime which may be problematic. Since this needs to be injected into the ctor of the index, all of these things should be injected into the ctor of this object too. This is a new object so we don't need to work around breaking changes and introduce spaghetti code. I feel like this whole class can be simplified, it takes in it's requirements in the ctor and exposes them by properties (not methods).

Reply:
Any thoughts on this?

Reply:
@Shazwazza I am going to look into that soon, as need figure out maybe better way of handling as I also need introduce background queue and need figure out best way of handling that :)